### PR TITLE
improve user-friendliness and add support for linux/1password

### PR DESCRIPTION
**Pull request by sukoneck**
names the tool `AWS CLI Helper`

adds support for:
- awsup for quick start
- 1password as an OTP engine
- plaintext file as vault engine
- linux

### CODE REVIEW

File: awsup.sh [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+#
+#
+
+set -eo pipefail
+echo "🆙 Starting awsup..."
+
+# Set the target directory
+TARGET_DIR="${HOME}/.aws"
+REPO_URL="https://github.com/0x416e746f6e/.aws.git"
+REPO_BRANCH="main"
```
Comment (sukoneck) [inline awsup.sh:12]: note:

to test prior to merging you'll need to manually update these.

File: awsup.sh [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+# Check if the config or credentials files already exist
+CONFIG_FILE="${TARGET_DIR}/config"
+CREDENTIALS_FILE="${TARGET_DIR}/credentials"
+
+# Get the current date for the directory names
+DATE=$(date +%Y-%m-%d_%H-%M-%S)
+BACKUP_DIR="${TARGET_DIR}/backup-${DATE}"
+TEMP_DIR="$( pwd )/awsup-${DATE}"
```
Comment (0x416e746f6e) [inline awsup.sh]: **suggestion:**

```suggestion
TEMP_DIR="$( mktemp -d -t awsup-XXX )"
```

File: awsup.sh [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+    # Only backup the files, not subdirectories
+    for item in "${TARGET_DIR}"/*; do
+        if [ -f "${item}" ]; then
+            mv "${item}" "${BACKUP_DIR}"
+        fi
+    done
+}
+
+handle_file_backup
```
Comment (0x416e746f6e) [inline awsup.sh]: **suggestion:**

it's easier to follow the flow if function declarations are grouped separately from their respective invocations.  otherwise, this statement for example is easy to overlook as it "hides" between functions (it's not so big of an issue when the functions are small, but they tend to grow over time).

File: login.sh
Comment (0x416e746f6e) [inline login.sh:1]: **issue:**

I am using this file very heavily in my workflow.

File: lib.sh [hunk 4]
```diff
@@ -1,88 +1,356 @@
...
+      ;;
+  esac
+}
+
+function get_1fa_token_osxkeychain() {
   security find-generic-password \
-    -l "$AWS_USER_ARN" \
-    -a "$AWS_USER_ARN" \
-    -s "$AWS_USER_ARN" \
+    -l "${AWS_HELPER_USER_ARN}" \
```
Comment (0x416e746f6e) [inline lib.sh:20]: **question:**

where does this (env) var come from?

the only place I see is the _local_ var definitions in `save_1fa_token_xxx` functions. however those vars are _local_, so should not be visible outside `save_1fa_token_xxx` scopes.

File: lib.sh [hunk 5]
```diff
@@ -1,88 +1,356 @@
-AWS_MFA_DEVICE="arn:aws:iam::${AWS_ACCOUNT_ID}:mfa/${AWS_IAM_USERNAME}"
-AWS_USER_ARN="arn:aws:iam::${AWS_ACCOUNT_ID}:user/${AWS_IAM_USERNAME}"
+AWS_HELPER_PROFILE="${HOME}/.profile"
+
+function source_profile() {
```
Comment (0x416e746f6e) [inline lib.sh]: **question:**

why is this necessary?

sourcing of `~/.profile` or `~/.bash_profile` is the job of the shell itself, not of this helper.

File: awsup.sh (hunk 2 above)
Comment (sukoneck) [reply awsup.sh]: great idea!

File: lib.sh (hunk 4 above)
Comment (sukoneck) [reply lib.sh:20]: anything persistent is being prompted for in the `save_setup()` [function here](https://github.com/0x416e746f6e/.aws/pull/3/files?diff=split&w=0#diff-503615c901c45226b03b60450608983b94776c2b0267f89ca6ef20429007cc66R288) and then saved in `~/.profile`. 

the user is prompted to run `save_setup()` when [awsup finishes](https://github.com/0x416e746f6e/.aws/pull/3/files?diff=split&w=0#diff-503615c901c45226b03b60450608983b94776c2b0267f89ca6ef20429007cc66R288).

File: awsup.sh (hunk 3 above)
Comment (sukoneck) [reply awsup.sh]: oops! agree, fixed.

File: lib.sh (hunk 5 above)
Comment (sukoneck) [reply lib.sh]: in a previous iteration I needed it. agreed, removed.

File: login.sh
Comment (sukoneck) [reply login.sh:1]: I added it back in. currently you can switch over by replacing `login.sh` with `helper.sh --login`. 

is there another way you'd prefer to handle this?

**Comment by sukoneck** [comment]
replaced by https://github.com/flashbots/.aws/pull/1
